src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -894,41 +894,54 @@ private Task DecodeTrailersAsync(bool endHeaders, in ReadOnlySequence<byte> payl
 
         private void StartStream()
         {
-            if (!_isMethodConnect && (_parsedPseudoHeaderFields & _mandatoryRequestPseudoHeaderFields) != _mandatoryRequestPseudoHeaderFields)
-            {
-                // All HTTP/2 requests MUST include exactly one valid value for the :method, :scheme, and :path pseudo-header
-                // fields, unless it is a CONNECT request (Section 8.3). An HTTP request that omits mandatory pseudo-header
-                // fields is malformed (Section 8.1.2.6).
-                throw new Http2StreamErrorException(_currentHeadersStream.StreamId, CoreStrings.Http2ErrorMissingMandatoryPseudoHeaderFields, Http2ErrorCode.PROTOCOL_ERROR);
-            }
+            // The stream now exists and must be tracked and drained even if Http2StreamErrorException is thrown before dispatching to the application.
+            _streams[_incomingFrame.StreamId] = _currentHeadersStream;
```

Comment:
Is there a limit on this? Could a malicious client start a huge number of streams and force the server to track them all?

Reply:
We have the _serverSettings.MaxConcurrentStreams and _serverSettings.MaxConcurrentStreams * 2 limits below, but you're right, we will just keep adding to this collection. Maybe we should have these three tiers:
- MaxConcurrentStreams -> Refused
- MaxConcurrentStreams * 1.5 -> ENHANCE_YOUR_CALM
- MaxConcurrentStreams * 2 -> Connection error.

Reply:
After talking to @halter73 here's the modified plan:
- MaxConcurrentStreams -> Refused, drain queue
- MaxConcurrentStreams * 2 -> ENHANCE_YOUR_CALM, drain queue
- If the drain queue is full, drop the oldest item from the queue.

This should allow the gRPC stress scenario to start as many streams as it likes, have many refused, and avoid closing the connection so long as the data frames for the refused streams arrive before too many more new streams. E.g. Headers1, Data1, Headers3, Data3 should always be safe.